aiida/engine/processes/builder.py

```diff
@@ -7,10 +7,11 @@
 # For further information on the license, see the LICENSE.txt file        #
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
-# pylint: disable=cell-var-from-loop
+# pylint: disable=cell-var-from-loop, line-too-long
```

Comment:
I wouldn't disable this.

Reply:
I agree. I have to say the `cell-var-from-loop` is also probably not a good idea to ignore. I know it was already there, but this might point at a bug

Reply:
When enabling the `cell-var-from-loop` check, it complains about lines 52 (`port`) and 60 (`name`).

https://github.com/aiidateam/aiida-core/blob/9244969fe53b16a89309624b32a50b0c6b3ee768/aiida/engine/processes/builder.py#L41-L65

To me, the first looks like a false positive, while the second could be a bug. Here's some roughly-equivalent test code:
```python
In [4]: funcs = []

In [5]: mapping = dict()

In [6]: for x in range(10):
   ...:     def foo(val=x):
   ...:         mapping[x] = val
   ...:         return val
   ...:     funcs.append(foo)
   ...:
   ...:

In [7]: [f() for f in funcs]
Out[7]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [8]: mapping
Out[8]: {9: 9}
```

What's puzzling to me is that this looks like quite often-used code. Maybe both are actually false positives. But it certainly warrants looking into; `cell-var-from-loop` is quite a useful check in my experience, since it covers easily-overlooked corner cases of the language.

Reply:
1. Re `cell-var-from-loop`, I'm not sure I fully understand the problem (it looks somewhat simililar to the recursively expanded and simply expanded variables that were such a headache when coding makefiles).

2. Re `line-too-long`, I've managed most of them but there are two that are still there and I don't know how to fix without introducing intermediate variables (all "aesthetical" modifications are reverted by yapf):

```
    def __dir__(self):
        return sorted(set(self._valid_fields + [key for key, _ in self.__dict__.items() if key.startswith('_')]))
(...)
    if not (isinstance(pruned, collections.abc.Mapping) and not pruned and not isinstance(pruned, Node)):
        result[key] = pruned
```

Reply:
@sphuber I'm seriously confused by that code. Specifically, the `setattr(self.__class__, name, getter)` sets the property _on the `ProcessBuilderNamespace` class_ - but doesn't that mean all `ProcessBuilderNamespace` instances will share the same properties?

Also, when trying to reproduce the class behavior, I can't get the `fsetter` to work - instead, the `__setattr__` is needed. But why even have the `fsetter`, then?

The following code might illustrate my confusion somewhat:

```python
class Foo:
    def __init__(self, num_vals, obj_name):
        self._data = {}
        for x in range(num_vals):
            name = chr(97 + x)

            def fgetter(self, name=name, obj_name=obj_name):
                print(f'fgetter for {obj_name}')
                return self._data.get(name)

            def fsetter(self, value, obj_name=obj_name):
                print(f'fsetter for {obj_name}')
                self._data[name] = value

            getter = property(fgetter)
            getter.setter(fsetter)

            setattr(self.__class__, name, getter)

    def __setattr__(self, attr, value):
        if attr.startswith('_'):
            object.__setattr__(self, attr, value)
        else:
            self._data[attr] = value


f = Foo(3, 'f')
print(f.a)
f.a = 2
print(f.a)
try:
    print(f.j)
except AttributeError as exc:
    print(str(exc))

g = Foo(10, 'g')
print(g.j)
print(f.j)
```
Output:
```
fgetter for f
None
fgetter for f
2
'Foo' object has no attribute 'j'
fgetter for g
None
fgetter for g
None
```

Reply:
> Re cell-var-from-loop, I'm not sure I fully understand the problem (it looks somewhat simililar to the recursively expanded and simply expanded variables that were such a headache when coding makefiles).

@ramirezfranciscof  Right, so this is bit of an obscure part of Python, it touches on how _function closure_ works. When you define

```python
x = 1
def foo():
    print(x)
```
how is `x` "passed" into `foo`? In the above example it doesn't matter, but what about

```python
funcs = []
for x in range(10):
    def foo():
         print(x)
    funcs.append(foo)

for foo in funcs:
    foo()
```

What would you expect that to print?

---

In short, the closure is done _by name_, and if that name is bound to a different value in the enclosing scope, calling `foo` gets the _changed_ `x`.
So yeah, that's quite a gotcha, and it's nice that `pylint` warns us about it.

Reply:
Looking at that code, I actually found a (quite serious IMO) bug, see https://github.com/aiidateam/aiida-core/issues/4420. It's not related directly to this change, though.

Reply:
@greschd Ah, I think I understand. When you use an externally defined variable inside the definition of a function, when you call the function it will use whatever is stored in that variable at the time it evaluates the function. We are doing a subtle version of this, not by using the external variable inside the function, but using it to set a default.

I am now wondering what is the proper way to "automatically generate parametrized functions". For example, I want to have a list of 100 functions, where the i-function will add i to the number passed. I would have tried the exact thing in your example:

```python
funcs = []
for x in range(100):
    def foo(n):
         return n+x
    funcs.append(foo)
```

But then this doesn't work...how would you do this then?

Reply:
> We are doing a subtle version of this, not by using the external variable inside the function, but using it to set a default.

When setting the default, it actually binds to the _value_ of that variable at the time of _defining_ the function. See the following:

```python
funcs = []
for x in range(10):
    def foo(y=x):
         print('x', x)
         print('y', y)
    funcs.append(foo)

for foo in funcs:
    foo()
```

This is the "false positive", which I think is this `pylint` issue: https://github.com/PyCQA/pylint/issues/3107

The "true positive" in our code is this line:
https://github.com/aiidateam/aiida-core/blob/9244969fe53b16a89309624b32a50b0c6b3ee768/aiida/engine/processes/builder.py#L60

Reply:
> The "true positive" in our code is this line:

AH, because it is interpreting `name` as the external variable instead of `name` the internal function parameter?

> This is the "false positive", which I think is this pylint issue: PyCQA/pylint#3107. And incidentally, this is also how you can circumvent the problem.

Do you mean, in my example you would have to pass x as the default of an "unused/artificial/forced" function parameter in order to do my list of summing functions?


```python
funcs = []
for x in range(100):
    def foo(n, int_x=x):
         return n + int_x
    funcs.append(foo)
```

Mmmm It feels kind of hacky (the summing term is an internal parameter, I don't want to have to define an input that the user could potentially overwrite).

Reply:
> 
> 
> > The "true positive" in our code is this line:
> 
> AH, because it is interpreting `name` as the external variable instead of `name` the internal function input parameter?
> 

Right, because that function doesn't _have_ a `name` parameter, it's just "captured" from the enclosing scope.

> 
> Do you mean, in my example you would have to pass x as the default of an "unused/artificial/forced" function parameter in order to do my list of summing functions?
> 
> ```python
> funcs = []
> for x in range(100):
>     def foo(n, int_x=x):
>          return n + int_x
>     funcs.append(foo)
> ```
> 
> Mmmm It feels kind of hacky (the summing term is an internal parameter, I don't want to have to define an input that the user could potentially overwrite).

Yep, that'd work. If you want to avoid exposing the extra keyword argument but still bind it to the object, you can use `functools.partial`, like so:

```python
In [1]: from functools import partial

In [4]: funcs = []
   ...: for x in range(10):
   ...:     def foo(y):
   ...:         return y
   ...:     funcs.append(partial(foo, y=x))
   ...:
```

Note also how you wouldn't even have to re-define `foo` in every loop iteration here:
```python
In [6]: funcs = []
   ...: def foo(y):
   ...:     return y
   ...: for x in range(10):
   ...:     funcs.append(partial(foo, y=x))
   ...:
```

For the general behavior, there's also an entry in the Python FAQ: https://docs.python.org/3/faq/programming.html#why-do-lambdas-defined-in-a-loop-with-different-values-all-return-the-same-result

Reply:
And here's the somewhat terse description of name binding, which is the background for _why_ it's behaving that way: https://docs.python.org/3.8/reference/executionmodel.html#naming-and-binding

Reply:
> Right, because that function doesn't have a name parameter, it's just "captured" from the enclosing scope.

I see. I was mixing up `fsetter` with `fgetter`, but you are right, the error comes from the one with no `name` parameter. Ok, I think the solution is then to copy the def just above and add the `name=name` thing, right? And add the ignore in the specific line for that other function.

Thanks for the thorough explanation, btw! Crystal clear!

Reply:
> Ok, I think the solution is then to copy the def just above and add the `name=name` thing, right?

In principle yes, but I think the rest of that `__init__` code doesn't do exactly what it's supposed to - which is why I opened #4420. I'm not 100% sure, but I think the `fsetter` is never actually invoked. 

But I think it would be ok to fix it in this way here (for the false positive, just explicitly ignore the pylint error _on that line_). Resolving #4420 is probably not closely related to this PR - I just happened to stumble upon it while looking at this.

> Thanks for the thorough explanation, btw! Crystal clear!

Happy to help!

Reply:
Yeah, I just did the modifications. This way you will also be able to get "cleaner" base for the test of #4420 .